### PR TITLE
Cleanup the codebase

### PR DESCRIPTION
**Pull request by ItsDrike**
There are many older code remnants scattered in the project, several non-pythonic implementations for certain things, and approaches which involve too much repetition or have other issues,

This PR is an attempt to clean up the code-base a bit to make it more readable and maintainable.

### CODE REVIEW

**Comment by ItsDrike** [comment]
As a part of this PR, I also wanted to convert at least most bare exceptions, or `except Exception` statements to only except specific exceptions explicitly, since bare exception will also capture things like keyboardinterrupts, etc. If this is desired, it should be marked, also, it's a place where a lot of bugs may get captured without any feedback of what happened, if we need to catch some exception in a block of code, it should always be explicit.

To do this, I've handled (or will handle) some of these from what I figured they were supposed to be capturing, so when reviewing please take in mind that these were mostly my assumptions on what these try-except blocks were handling and it may very well be a wrong assumption, so please pay close attention to these to avoid this PR stopping these from handling some errors which should've been captured, but also make sure I haven't used a way too general exception (for example capturing every error belonging to certain library, when only one or a few of those would suffice).

However apart from these, there were some try-except blocks which I found in the code-base, but wasn't able to identify which exceptions exactly are they supposed to catch. This is the list of those:

- https://github.com/Dinnerbone/mcstatus/blob/a31e808b1f35190c3875645e40d5bc31a6f3a950/mcstatus/scripts/mcstatus.py#L80-L102
- https://github.com/Dinnerbone/mcstatus/blob/a31e808b1f35190c3875645e40d5bc31a6f3a950/mcstatus/protocol/connection.py#L211-L215 https://github.com/Dinnerbone/mcstatus/blob/a31e808b1f35190c3875645e40d5bc31a6f3a950/mcstatus/protocol/connection.py#L248-L252 https://github.com/Dinnerbone/mcstatus/blob/a31e808b1f35190c3875645e40d5bc31a6f3a950/mcstatus/protocol/connection.py#L280-L284 https://github.com/Dinnerbone/mcstatus/blob/a31e808b1f35190c3875645e40d5bc31a6f3a950/mcstatus/protocol/connection.py#L317-L321

File: mcstatus/pinger.py
Comment (kevinkjt2000) [inline mcstatus/pinger.py:240]: Were there some test cases that you used to test this? It should be straightforward to test a few dictionary inputs in the unit test. Would you mind adding those?

File: mcstatus/protocol/connection.py
Comment (kevinkjt2000) [inline mcstatus/protocol/connection.py:262]: `typing.Optional` exists for cases like this
```suggestion
    reader: Optional[asyncio.StreamReader] = None
    writer: Optional[asyncio.StreamWriter] = None
    timeout: Optional[float] = None
```

File: mcstatus/protocol/connection.py
Comment (kevinkjt2000) [inline mcstatus/protocol/connection.py:294]:
```suggestion
    stream: Optional[asyncio_dgram.aio.DatagramClient] = None
    timeout: Optional[float] = None
```

File: mcstatus/scripts/mcstatus.py
Comment (kevinkjt2000) [inline mcstatus/scripts/mcstatus.py:7]:
```suggestion
server: Optional[MinecraftServer] = None
```

**Review by kevinkjt2000 — CHANGES_REQUESTED** [review]
Very nice! I'll happily accept this after suggested changes that I've made via comments and a PR.

File: mcstatus/protocol/connection.py
Comment (ItsDrike) [reply mcstatus/protocol/connection.py:262]: I'm aware of that, however there was a reason I did this instead. I wanted to avoid further type errors down the line as if these types were defined as typing.Optional, every time you'd do for example `reader.x` the type checker would complain that `NoneType` has no attribute `x`.

The way around this while still passing type checkers would be to add these kinds of type ignore lines everywhere these variables will actually be used and ignore these, however I think that's a lot harder to maintain and it's just more unnecessary code clutter with these ignore comments. An alternative approach would be type narrowing by having a condition that checks if the type is none and raises an exception/returns prematurely (or having the logic in an exception that ensures the type isn't none). That would look like this:
```py
if reader is None:
    raise SomeException
``` 

Should I really still make this optional? If yes, which of these approaches should I take?

Comment (kevinkjt2000) [reply mcstatus/protocol/connection.py:262]: pytype passed when I made this modification locally. Are you sure this would cause an issue for type checking?

Comment (ItsDrike) [reply mcstatus/protocol/connection.py:262]: It is a typing issue, and both pyright and mypy do catch it as such, I'm not sure why pytype wouldn't be catching this but it's certainly a typing issue. The type checkers can't know that a function which sets these values is expected to run after initialization, if that's desired, it should be in init, not in a separate function. Because of that type-checkers can't narrow that type down and drop the default `None`, and obviously, accessing an attribute that only one of the union types have is a typing issue.

Minimal reproducable example:
```py
from typing import Union

x: Union[dict, list] = [1, 2, 3]

# This obviously can't work, `dict` doesn't support `append`
# it's the same error as None not having some func/attribute while the other type does
# it simply can't be right type-wise
x.append(5)
```

File: mcstatus/pinger.py
Comment (ItsDrike) [reply mcstatus/pinger.py:240]: Done in c3ebf12, (which actually did uncover an issue in the implementation, which I've then fixed in 
8cdcf6b).

However I think there may be an issue with this implementation. This does follow the same logic as the original implementation did so if this really was an issue, it was there originally too. The thing I'm talking about is for example something like this:
```py
        out = PingResponse._parse_description(
            {
                "extra": [
                    {
                        "bold": True,
                        "italic": True,
                        "color": "gray",
                        "text": "foo"
                    },
                    {
                        "color": "gold",
                        "text": "bar"
                    }
                ],
                "text": "."
            }
        )
        # We don't care in which order the style prefixes are, allow any
        assert out in {
            "§l§o§7foo§6bar.",
            "§l§7§ofoo§6bar.",
            "§o§l§7foo§6bar.",
            "§o§7§lfoo§6bar.",
            "§7§o§lfoo§6bar.",
            "§7§l§ofoo§6bar.",
        }
```
What I'm unsure about here is the fact that for example the ending `"text"` probably shouldn't have any style, yet we never reset the colors, same goes for the text in the second extra group, it probably shouldn't be affected by the style of the first one, but we never reset the color after going to another group. We just add on the style/colors we found in that next group, not worrying about any style resets, should we?

**Comment by ItsDrike** [comment]
Note about commit 9b35cef failing `pytype`:

For some reason, pytype doesn't play nicely with overriding type-hints directly in overridden init. After calling the init from superclass, it basically ignores this new type assignment. This causes pytype to report issues in multiple places as it thinks `self.connection` is `TCPSocketConnection` rather than `TCPAsyncSocketConnection`. 

I'm not sure how should I go about fixing this, both pyright and mypy are fine with this kind of assignment.

File: mcstatus/pinger.py
Comment (kevinkjt2000) [inline mcstatus/pinger.py:12]: These could be a bit complicated. §m and §r are missing.
https://minecraft.fandom.com/wiki/Formatting_codes

File: mcstatus/protocol/connection.py
Comment (kevinkjt2000) [reply mcstatus/protocol/connection.py:264]: `Optional[T]` is defined as `Union[T, type(None)]`, so what you are saying makes sense to me now
https://github.com/python/cpython/blob/7c5b01b5101923fc38274c491bd55239ee9f0416/Lib/typing.py#L363-L365

**Comment by kevinkjt2000** [comment]
I'm okay with ditching pytype if another tool works correctly. My experience with every python type checker has been that they all have bugs, pros, and cons (kinda like every other piece of software).

> I'd first like to get a response on what errors these try-except blocks are supposed to catch so that these can be changed to only catch those errors specifically. So when reviewing, I'd appreciate a comment explaining these before approving and merging.

I don't know what they were either, so of course, I'm going to approve and merge when tests pass :smile:

**Review by kevinkjt2000 — APPROVED** [review]
There's a small conflict that needs fixing and `tox format` would be nice, but you have my approval :+1:

File: mcstatus/protocol/connection.py
Comment (ItsDrike) [reply mcstatus/protocol/connection.py:296]: I assume this can also be marked as resolved following the discussion on why I didn't use optional above.

**Comment by ItsDrike** [comment]
> I'm okay with ditching pytype if another tool works correctly.

I'm willing to replace pytype with pyright, it should result in much faster type checking as pytype is really slow sometimes and I find it a lot more sensible in some places than pytype (at least from what I've seen in my very limited experience with it). Pyright is quite popular and supports most new typing features relatively quickly after they're added and it is maintained by Microsoft. 

An alternative would be mypy which is maintained by the python software foundation directly, however I often find mypy way too strict for certain things for no real reason which results in needing way too many ignores and overrides, it's also a bit slower than pyright, though still beats pytype by a long shot. There's also pyre-check and a few other less popular options, but I don't have any experience with those so I can't recommend them.

I don't think a cleanup PR is a place to change type checkers thogh, it may be fine to merge this first, even without passing typing requirements of pytype and fix that with another PR later, that's purely for switching type checkers. Alternatively this PR could simply wait until that PR gets in, though there are some typing fixes and this code does already pass pyright, if this were supposed to be done before this gets merged, it would be a bit annoying as to fulfill pyright many of the same things done here would need to happen in another PR again.

> I don't know what they were either, so of course, I'm going to approve and merge when tests pass 😄

Yeah, that's a problem I'd like to solve before this gets merged, wide try-except statements are often a huge problem when debugging, I'll try to find what they're for and replace at least most of them, though I can't promise I won't introduce any bugs, but that's probably still better than keeping it as is since those should get discovered fairly quickly and shouldn't be too hard to fix, and it will make the code a lot easier to debug afterwards.

**Additional note:**
I've left a comment in the resolved conversation on adding tests for the style description parsing pointing out a possible issue that may be occurring, since you marked as resolved, I assume it's not an issue? Though I'd like a more concrete response on that in case you missed it and just marked it as resolved since the test itself is now there.

**Comment by kevinkjt2000** [comment]
Yes, we could save fixing the typing stuff for later. Additionally I could delay creating a release.

Thanks for doing all this! I haven’t been able to donate a lot of time lately to this project.

File: mcstatus/pinger.py
Comment (kevinkjt2000) [reply mcstatus/pinger.py:240]: I'll open an issue to remind me to look at adding test cases for fancy statuses, but this refactoring that you have done should have the same behavior as the existing code. The while loop might be unnecessary complexity since it only goes one level in the "color" part of the style map, but I'm alright with that. It could be useful for nesting more types of styles in the future :man_shrugging:

Comment (kevinkjt2000) [reply mcstatus/pinger.py:240]: Do you have an example minecraft server that provides their MOTD/description back in this dictionary format? I tried this out on a vanilla 1.18.1 server with an MOTD of `\u00A7l\u00A77bold and gray\u00A7rreset` and the `.status().raw["description"]` had `§l§7bold and gray§rreset`. Which Minecraft servers return a dictionary here instead of the string?

Comment (ItsDrike) [reply mcstatus/pinger.py:240]: > IDo you have an example minecraft server that provides their MOTD/description back in this dictionary format?

It took me a while to find one too, but `2b2t.org` did return a dict like this.

**Pull request merged**
